### DISCUSSION TITLE
System.ServiceModel.Channels.Message and CoreWCF.Message incompatibility issue. Any solution for this?

### DISCUSSION THREAD

**Discussion by rohit99dev** [CoreWCF/CoreWCF#1470] [Q&A] [answered]
I have created a CoreWCF service that runs perfectly fine when hosted. When I try to consume it into another CoreWCF service, It throws below error:

```
<?xml version="1.0" encoding="utf-16"?>
<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/">
  <s:Header />
  <s:Body>
    <s:Fault>
      <faultcode xmlns:a="http://schemas.microsoft.com/ws/2005/05/addressing/none">a:ActionNotSupported</faultcode>
      <faultstring xml:lang="en-GB">The message with Action 'http:xxxxxxx/Request'
	  cannot be processed at the receiver, 
	  due to a ContractFilter mismatch at the EndpointDispatcher.
	  This may be because of either a contract mismatch
	  (mismatched Actions between sender and receiver) 
	  or a binding/security mismatch between the sender and the receiver. 
	  Check that sender and receiver have the same contract and the same binding
	  (
including security requirements, e.g. Message, Transport, None).</faultstring>
    </s:Fault>
  </s:Body>
</s:Envelope>
```
**Hosted - Service contracts**
```
using System.Threading.Tasks;
using CoreWCF;
using CoreWCF.Channels;

namespace FraudChecker
{
    [ServiceContract]
    public interface IFraudMessageChecker
    {
        [OperationContract(Action = "http://xxxxxx/xxxx"
       , ReplyAction = "http://xxxxxx/xxxx")]
        Task<Message> CheckRisk(Message checkRiskRequest);
    }
}
```
**Client consuming above service implementation:**
```
public interface IFraudMessageChecker
{
    
    [System.ServiceModel.OperationContractAttribute(Action="http://xxxxx/xxxxx", ReplyAction="http://xxxx/xxxx")]
    System.ServiceModel.Channels.Message CheckRisk(System.ServiceModel.Channels.Message request);
    
    [System.ServiceModel.OperationContractAttribute(Action="http://xxxxx/xxxxx", ReplyAction="http://xxxx/xxxx")]
    System.Threading.Tasks.Task<System.ServiceModel.Channels.Message> CheckRiskAsync(System.ServiceModel.Channels.Message request);
}
```
I think it's because the server has **corewcf.Message** attributes defined and client can only use **system.servicemodel.channel.Message** because of **System.ServiceModel.ClientBase**. I am wondering what would be the best possible solution or recommendation to fix this? Thank you!

**Comment by rohit99dev** [comment]
@mconnew, any help would be much appreciated. Thank you!

**Reply by mconnew** [reply]
The client must use System.ServiceModel attributes. The service can use either CoreWCF attributes, or System.ServiceModel attributes. While convenient to use the same attributes on the service side, it does require you to pull in the System.ServiceModel.Primitives client pacakge into your service project. This only extends to attributes though and doesn't apply to operation parameters or return types. In your case, you are requesting the raw Message object be passed. On the service side you must use CoreWCF.Message as the parameter and return type. Btw, on the service side, if you have a method in your contract which takes a single Message parameter, and returns a Message, it's possible to set the Action to "*" and it will accept any message. This might not be useful as you need to check the Action from the message headers yourself, and need to manually set the Action header on the reply message too, but for some situations, it can be useful. One scenario it can be useful for is to do this:
```csharp
    [ServiceContract]
    public interface IFraudMessageChecker
    {
        [OperationContract(Action = "http://xxxxxx/xxxx", ReplyAction = "http://xxxxxx/xxxx")]
        Task<Message> CheckRisk(Message checkRiskRequest);
        [OperationContract(Action = "*")
        Task<Message> UnhandledOperation(Message message);
    }
```
And then you can see what's wrong with the message. You can call ToString() and it will output a text XML representation of the message. Note, this does NOT output what is sent on the wire, it's a reconstruction of what's semantically equivalent. E.g. it will output an Action header in the SOAP message headers, but for SOAP1.1, this is actually sent as an HTTP header.  
Let me know if you still need some more pointers after this.

**Reply by mconnew** [reply]
I want to clarify a point, do you have two services, ServiceA and ServiceB, and ServiceA receives a request and you are trying to pass the incoming Message object to ServiceB, basically proxying the call? If so, you need to convert the incoming CoreWCF.Message object into a System.ServiceModel.Message object. I've already written code to do this in our unit tests as I created a WCF Client binding which uses System.ServiceModel.Message which directly dispatches to CoreWCF without doing any IO (which needs CoreWCF.Message), so I had to write conversion code to go back and forth between them. You can copy that code into your own project. The two static methods you will need [are here](https://github.com/CoreWCF/CoreWCF/blob/main/src/CoreWCF.Primitives/tests/Helpers/TestHelper.cs#L37-L51), and you can find the rest of the helper methods needed called from there.

**Reply by mconnew** [reply]
Btw, that code isn't the fastest, I want to be able to debug it easily. Look for where TextMessageEncodingBindingElement is used and replace it with BinaryMessageEncodingBindingElement to get a faster conversion if you are performance sensitive and your messages are non-trivial in size. If the messages are small, it likely won't make any difference and having the ability to be able to read as text the intermittent buffer can be useful if you're debugging an issue.

**Reply by rohit99dev** [reply]
Thanks for your reply @mconnew . I have three existing WCF services, all are migrated to dotnet Core using CoreWCF. 
**Service A** consumes **Service B** and as you said Service A receives a request and pass down the incoming Message object to ServiceB. Since Service B is also CoreWCF now, it expects Message of type CoreWCF.Bindings. I'll try your unit test code, it looks promising to me. My input request is generally around 400-500 lines of xml and is very performance sensitive.

**Reply by mconnew** [reply]
Did you mean Bindings or Message?

**Reply by rohit99dev** [reply]
Sorry, I meant Message

**Reply by mconnew** [reply]
No worries. Just checking because there are some types you wouldn't expect to send as part of a request which are serializable in .NET Framework that we haven't put in the effort to make sure they are similarly serializable in .NET and I wanted to make sure that you weren't depending on something like that. They are used for things like WS-Discovery (not currently supported) which needs to provide information about an endpoint over a SOAP response, so checking you aren't working in that problem space.

**Reply by rohit99dev** [reply]
@mconnew Need help with understanding the best place to use TestHelper.cs Method you shared above? 
Re-iterating structure: 

- **Service A Contract (CoreWCF.Channels.Message)**--> 
```
[OperationContract(Action = ""http://xxxxxx/xxxx""
        , ReplyAction = ""http://xxxxxx/xxxx"")]
[FaultContract(typeof(FraudCheckOrchestratorFault))]
Task<Message> CheckRisk(Message checkRiskRequest);
```
- **Service B Contract (CoreWCF.Channels.Message)**-->
```
[ServiceContract]
    public interface IFraudMessageChecker
    {
        [OperationContract(Action = "http://xxxxxx/xxxx"
       , ReplyAction = "http://xxxxxx/xxxx")]
        Task<Message> CheckRisk(Message checkRiskRequest);
    }
```
- **CoreWCF Service A consumes Service B**
**Reference.cs**
```
 public System.Threading.Tasks.Task<System.ServiceModel.Channels.Message> CheckRiskAsync(System.ServiceModel.Channels.Message request)
 {
     return base.Channel.CheckRiskAsync(request);
 }
```

**I tried converting incoming request as you recommended in Service A to System.ServiceModel.Channels.Message and pass it to Service B but throws same error as below. Is there any change required at Service B as well?**
```
<faultstring xml:lang="en-GB">The message with Action 'http://xxx/xxxx/xxx' cannot be processed at the receiver, 
	  due to a ContractFilter mismatch at the EndpointDispatcher.
	  This may be because of either a contract mismatch (mismatched Actions between sender and receiver) or 
	  a binding/security mismatch between the sender and the receiver.  
	  Check that sender and receiver have the same contract and the same binding
	  (including security requirements, e.g. Message, Transport, None).</faultstring>
    </s:Fault>
```

**Reply by mconnew** [reply, marked as answer]
On the service side of your app you always need to be referencing the CoreWCF.Channels.Message type. So the interface you specify in your AddServiceEndpoint call must have the operations reference CoreWCF.Channels.Message, and the implementation you pass to AddService\<TService\> will implement the contract interface and also use CoreWCF.Channels.Message for its parameters.  
When you instantiate a ChannelFactory\<TChannel\>, the interface type you specify for TChannel must use System.ServiceModel.Channels.Message in its operations. Once you've got that all worked out, your service implementation would look like this:
```c#
public class ServiceA : IServiceA
{
    public async Task<CoreWCF.Channels.Message> CheckRiskAsync(CoreWCF.Channels.Message request)
    {
        // CoreWCF deserializes the incoming SOAP request as a CoreWCF.Channels.Message
        // Convert that to a System.ServiceModel.Channels.Message
        System.ServiceModel.Channels.Message smRequest = TestHelper.ConvertMessage(request);
        // System.ServiceModel created client serializes a System.ServiceModel.Channels.Message into a SOAP message to send
        // The client will deserialize the reply SOAP message into a System.ServiceModel.Channels.Message
        System.ServiceModel.Channels.Message smReply = await base.Channel.CheckRiskAsync(smRequest);
        // Need to await the call to get the reply message
        // Then we convert it into a CoreWCF.Channels.Message so that CoreWCF can serialize that into a SOAP reply
        CoreWCF.Channels.Message reply = TestHelper.ConvertMessage(smReply);
        return reply;
    }
```

Hopefully this clears things up for you.

**Reply by rohit99dev** [reply]
@mconnew , Thankyou for your quick reply.  I have implemented the suggested way already in my code but unfortutanely it throws **ContractFilter mismatch exception** again. I have tried all possible ways but error does not seems to go away when calling base.Channel.CheckRiskAsync(request). My implementation looks like this:

**Service A implementation:**

```
public class ServiceA : IServiceA
{
    public async Task<CoreWCF.Channels.Message> CheckRiskData(CoreWCF.Channels.Message checkRiskRequest)
    {
        var messageProcessor = new MessageProcessor();
        await messageProcessor.Initialise(checkRiskRequest);
        
        using (var xmlreader = XmlReader.Create(new System.IO.StringReader(messageProcessor.GetBaseMessage().OuterXml)))
        {
            //Have tried both - System.ServiceModel.Create Message and CoreWCF
            var inboundMessage = CoreWCF.Channels.Message.CreateMessage(CoreWCF.Channels.MessageVersion.Soap11,
                "http://xxxx/xxxxx/CheckRiskRequest", xmlreader);
    
            System.ServiceModel.Channels.Message smRequest = ConvertMessage(inboundMessage);
            //**************FAILS ON BELOW LINE AT CHECKRISKASYNC()********************
            System.ServiceModel.Channels.Message smReply = await _fraudMessageChecker.CheckRiskAsync(smRequest);
            CoreWCF.Channels.Message reply = ConvertMessage(smRequest);
    
        }
    }
}
```


**Consumed Service B reference code in Service A**
```
public interface IFraudMessageChecker
{     
   [System.ServiceModel.OperationContractAttribute(Action="http://xxxx", ReplyAction="xxx")]
   System.Threading.Tasks.Task<System.ServiceModel.Channels.Message> CheckRiskAsync(System.ServiceModel.Channels.Message request);
}
public partial class FraudMessageCheckerClient : System.ServiceModel.ClientBase<FraudMessageChecker.IFraudMessageChecker>, FraudMessageChecker.IFraudMessageChecker
{        
     public System.Threading.Tasks.Task<System.ServiceModel.Channels.Message> CheckRiskAsync(System.ServiceModel.Channels.Message request)
      {
          //**Fails here**
          return base.Channel.CheckRiskAsync(request);
      }
}
```
**Error:**
```
 <faultstring xml:lang="en-GB">The message with Action 'http://xxx/xxx' cannot be processed at the receiver, 
	  due to a ContractFilter mismatch at the EndpointDispatcher. This may be because of either a contract mismatch
	  (mismatched Actions between sender and receiver) or a binding/security mismatch between the sender and the receiver. 
	  Check that sender and receiver have the same contract and the same binding (including security requirements, e.g. Message, Transport, None).</faultstring>
```

**Reply by rohit99dev** [reply]
@mconnew I have drilled down to the **issue it looks something to do with the operation contracts - Action and ReplyAction**. I tried a sample basic project implementing the same functionality and was throwing me the same error. But when I set Action and ReplyAction as *  i.e  ` [OperationContract(Action = "*", ReplyAction = "*")] `in service A and as well as B and It started working. Unfortunately, I can't do that in my main project so looking for the best solution.

**Reply by mconnew** [reply]
You might be needing to set the action manually. You can do that by setting smRequest.Headers.Action to what you need it to be, and then fixing the reply message in a similar way.

**Reply by mconnew** [reply]
@rohit99dev, if that doesn't work, you'll need to provide a simple repro solution demonstrating what you are trying to do. A repro just needs a single operation in each contract, doesn't need to be your full scenario.

**Reply by rohit99dev** [reply]
Thanks @mconnew , the issue is fixed now with the helper classes your shared :-)

**Reply by mconnew** [reply]
I'm curious. What was the thing which was blocking you? It might help someone else in the future.

**Reply by rohit99dev** [reply]
I added the header action of Service B in Service A when calling Service B. It fixed it.
```
//In service A
smRequest.Headers.Action = "http://ServiceB/response"
```

You recommended earlier to use BinaryMessageEncodingBindingElement instead of TestMessageEncoding if service is performance oriented. While after updating the method in helper class to Binary I get below error. Do I need to change incoming message to SOAP12 or is there inbuilt fix to it?

System.InvalidOperationException: 'The System.ServiceModel.Channels.BinaryMessageEncodingBindingElement binding element requires envelope version 'Soap12 (http://www.w3.org/2003/05/soap-envelope)' It doesn't support 'Soap11 (http://schemas.xmlsoap.org/soap/envelope/)'.'
